urlbd77b745af_KokaKiwi_MCLauncher_tgz-pJ8-net_contrapunctus_lzma_BenchmarkJ8/src/com/kokakiwi/mclauncher/core/updater/FileExtractor.java

```diff
@@ -0,0 +1,218 @@
+        
+        int ret = in.read(buffer);
+        while (ret >= 1)
+        {
+            out.write(buffer, 0, ret);
+            ret = in.read(buffer);
+        }
+        
+        in.close();
+        out.close();
+        
+        in = null;
+        out = null;
+        
+        return uncompressed;
+    }
+    
+    @SuppressWarnings("unchecked")
+    public static void extractZIP(LauncherAPI api, GameFile source, File file,
+            File dest, int min, int max) throws Exception
+    {
+        final ZipFile zip = new ZipFile(file);
+        
+        Enumeration<ZipEntry> entries = (Enumeration<ZipEntry>) zip.entries();
+        
+        int total = 0;
+        while (entries.hasMoreElements())
+        {
+            final ZipEntry entry = entries.nextElement();
+            
+            if (!entry.isDirectory())
+            {
+                total += entry.getSize();
+            }
+        }
+        
+        entries = (Enumeration<ZipEntry>) zip.entries();
+        final int current = 0;
+        
+        while (entries.hasMoreElements())
+        {
+            final ZipEntry entry = entries.nextElement();
+            
+            if (entry.isDirectory())
+            {
+                final File dir = new File(dest, entry.getName());
+                if (!dir.exists())
+                {
+                    dir.mkdirs();
+                }
+            }
+            else
+            {
+                final File e = new File(dest, entry.getName());
+                copyStream(api, source, zip.getInputStream(entry),
+                        new FileOutputStream(e), current, total, min, max);
+            }
+        }
+        
+        zip.close();
+    }
+    
+    public static void extractJAR(LauncherAPI api, GameFile source, File file,
+            File dest, int min, int max) throws Exception
+    {
+        JarFile jar = null;
+        try
+        {
+            jar = new JarFile(file);
+        }
+        catch (final IOException e)
+        {
+            e.printStackTrace();
+        }
+        
+        Enumeration<JarEntry> entries = jar.entries();
+        
+        int total = 0;
+        while (entries.hasMoreElements())
+        {
+            final JarEntry entry = entries.nextElement();
+            
+            if (!entry.isDirectory()
+                    && !entry.getName().equalsIgnoreCase("MANIFEST.MF"))
+            {
+                total += entry.getSize();
+            }
+        }
+        
+        entries = jar.entries();
+        final int current = 0;
+        
+        while (entries.hasMoreElements())
+        {
+            final JarEntry entry = entries.nextElement();
+            
+            if (entry.isDirectory())
+            {
+                final File dir = new File(dest, entry.getName());
+                if (!dir.exists()
+                        && !dir.getName().equalsIgnoreCase("META-INF"))
+                {
+                    dir.mkdirs();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Confirm that `mkdirs` was successful. Handle any failures that happened. If you fail silently, then you might encounter errors that are harder to debug. See the following diffs on GitHub that make this code change: [diff1](https://github.com/qunarcorp/qmq/commit/12f1d9d6a57a02bc2415cb2af9f4388a77b12c02#diff-6b67be6e17f24fdeddb7c725ac8ec5600d6e63be0bdca0963a83d530e6d55e35L36), [diff2](https://github.com/apache/jmeter/commit/1ec61206779b3f84df18516682b4791bf015178f#diff-cd74304763c097fbc085ac2bc952b46ab3843ac825cacacad89e8064296e955aL330)

Learn more: [File.mkdirs in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/io/File.html#mkdirs--)